### PR TITLE
SCHED-270: immutable decorator and suppression of deepcopy of immutable objects

### PR DESCRIPTION
**Pull request by sraaphorst**
In this PR:
* I added an `immutable` decorator, which redirects `copy` and `deepcopy` to return `self` instead of an actual copy of self.
* I cleaned up some dataclasses, to avoid using mutable members in hashing and removing all unsafe hashing, and marking members with `field` options to improve the code.
* Added gelidum library to freeze data structures that should be truly immutable (i.e. freezing lists and dicts instead of just their assigned references).
* Removed warnings with `# noqa` tag.
* Added a test case to test the `immutable` decorator and the results of `deepcopy`ing a `Program`.

### CODE REVIEW

File: lucupy/decorators/__init__.py [hunk 1]
```diff
@@ -18,3 +20,21 @@ def wrapper(*args, **kwargs):
         return np.squeeze(func(*args, **kwargs))
 
     return wrapper
+
+
+def immutable(cls):
```
Comment (stroncod) [inline lucupy/decorators/__init__.py:25]: One of us!

File: lucupy/minimodel/atom.py [hunk 2]
```diff
@@ -30,9 +30,9 @@ class Atom:
 
     """
     id: int
-    exec_time: timedelta
-    prog_time: timedelta
-    part_time: timedelta
+    exec_time: timedelta = field(hash=False, compare=False)
```
Comment (stroncod) [inline lucupy/minimodel/atom.py:33]: I assuming we only want to hash and compare by id.

**Review by stroncod — APPROVED** [review]
Great changes!

File: lucupy/decorators/__init__.py (hunk 1 above)
Comment (sraaphorst) [reply lucupy/decorators/__init__.py:25]: Ha! 🤖

File: lucupy/minimodel/atom.py (hunk 2 above)
Comment (sraaphorst) [reply lucupy/minimodel/atom.py:33]: I was thinking about that, too, and I think it makes sense. I don't think we ever actually use any of these as a key and only as a value, so we never hash on them, but I figured at least ensure the mutable parts are left out all the same.

**Pull request merged**